edb/edgeql/parser/grammar/convert.py

```diff
@@ -0,0 +1,418 @@
+                    replace_choice_of_options,
+                ]
+            )
+
+    return [
+        EBNF_Production(
+            production.name,
+            replace_inlines(production.item)
+        )
+        for production in productions
+    ]
+
+
+def to_ebnf(spec: parsing.Spec, path: pathlib.Path):
+    ebnf_productions: List[EBNF_Production] = []
+
+    # add token productions
+    for token in spec._tokens:
+        if token in ['<e>', '<$>']:
+            continue
+        ebnf_productions.append(EBNF_Production(token, EBNF_Literal(token)))
+
+    # add nonterm productions
+    nonterm_productions: Dict[str, List[List[EBNF_Reference]]] = {}
+
+    has_production = {
+        **{
+            token: True
+            for token in spec._tokens
+        },
+        **{
+            nonterm: False
+            for nonterm in spec._nonterms
+        },
+    }
+
+    for production in spec._productions:
+        prod_name = str(production.lhs)
+        if prod_name == '<S>':
+            continue
+
+        item_list: List[EBNF_Reference] = [
+            EBNF_Reference(item.name)
+            for item in production.rhs
+        ]
+
+        has_production[prod_name] = True
+
+        if prod_name not in nonterm_productions:
+            nonterm_productions[prod_name] = []
+        nonterm_productions[prod_name].append(item_list)
+
+    for prod_name, item_lists in nonterm_productions.items():
+        # if a production only refers to nonterminals with no productions
+        # remove it, since it does nothing
+
+        nonterm_productions[prod_name] = [
+            item_list
+            for item_list in item_lists
+            if
+                # refers to token or nonterm with productions
+                any(has_production[item.name] for item in item_list) or
+                # keep empty reductions, handled later as an optional
+                item_list == []
+        ]
+
+
+    for prod_name, item_lists in nonterm_productions.items():
+        if [] in item_lists:
+            # optional nonterm
+            item_lists = [
+                item_list
+                for item_list in item_lists
+                if item_list != []
+            ]
+            ebnf_productions.append(
+                EBNF_Production(prod_name, ebnf_single_or_choice([
+                    EBNF_Optional(ebnf_single_or_sequence(item_list))
+                    for item_list in item_lists
+                ]))
+            )
+        else:
+            ebnf_productions.append(
+                EBNF_Production(prod_name, ebnf_single_or_choice([
+                    ebnf_single_or_sequence(item_list)
+                    for item_list in item_lists
+                ]))
+            )
+
+    # repeatedly simplify until no simplifications are made
+    while True:
+        count = len(ebnf_productions)
+        ebnf_productions = simplify_productions(ebnf_productions)
+        if count == len(ebnf_productions):
+            break
+
+    # output
```

Comment:
I think the best place for this code to live would be as a tool in `edb/tools`. There are a bunch of examples there, and then we'd run it like `edb generate_ebnf` or some such.